### ISSUE TITLE
Introduce backends

### ISSUE THREAD

**Issue by soywod** [soywod/himalaya#295]
For now, we can only connect himalaya to an IMAP server with this config:

```
global-settings

[account-name]
account-settings
```

We want to connect other backends:

- [x] https://github.com/soywod/himalaya/pull/296
- [x] #43 
- [x] #57 

- [x] #318 

I propose to change the config this way:

```
global-settings

[backend-name.account-name]
account-settings
```

The fact to prefix the account with a backend allow us to have different APIs of account settings (one per backend). Plus it will match well with the toml to rust deserialization.

To avoid breaking changes, we can consider that the default backend (if no prefix given) is the IMAP one. But it could be nice to deprecate such implicit usage. To show a warning to those users, we will need https://github.com/soywod/himalaya/pull/293.

**Comment by matthiasbeyer** [comment]
Hi! For me as a user, the following scheme would be more logical

```
global-settings

[account.name]
backend = "<backend>"
```

**Comment by soywod** [comment]
It is what I thought first, but it is gonna be a pain to deserialize. Rust treats `[account]` section as a hash map, so you can have only one type of values. Backends will have different properties, so it is hardly compatible :disappointed:

**Comment by matthiasbeyer** [comment]
That's absolutely no problem with [serdes features (untagged enums)](https://serde.rs/enum-representations.html#untagged):

```
#[derive(serde::Deserialize)]
#[serde(untagged)]
enum Account {
    Imap(ImapAccount),
    Notmuch(NotmuchAccount)
    // ...
}
```

Or even better with [adjacently-tagged enums](https://serde.rs/enum-representations.html#adjacently-tagged), for example:

```
#[derive(serde::Deserialize)]
#[serde(tag = "backend", content = "configuration")]
enum Account {
    Imap(ImapAccount),
    Notmuch(NotmuchAccount)
    // ...
}
```

**Comment by soywod** [comment]
Oh, that's great! I did not know about these serde features. I will give a shot and propose a PR to switch to the new model. Then we will be able to do other PRs with backend implementations.

**Comment by soywod** [comment]
I have a working **ugly** POC https://github.com/soywod/himalaya/pull/296. I need to re-read again and improve it, but for now, here what I have:

- I use the untagged enum over the adjacently-tagged enum because I was not able to set up a default value for the adjacently-tagged enum (feel free to propose a solution if you have), which implies a breaking change (the `backend = "Imap"`).
- The POC is backward compatible, it works straight with the previous version of the config file
- I renamed the `ImapService` trait into `Backend`. So for now, the backend API is exactly the same as the IMAP, but we will need to adjust it depending on the other backend needs.

I let you know when the PR is ready for reviewing! Feel free to propose your opinion / ideas etc.

**Comment by soywod** [comment]
And thanks for your help :pray:

**Comment by soywod** [comment]
The PR is more or less ready for reviewing. Most of the changes are in `account_entity` and `config_entity`, the rest is just renaming `imap` into `backend`.

**Comment by soywod** [comment]
In parallel I try to think about the common backend interface. I start to write it down for later:

### Common to all

- connect
- read mailboxes
- read envelopes
- search envelopes
- read message
- add message
- add message flags
- set message flags
- remove message flags
- disconnect

### IMAP only

- watch / notify

### POP3 only
### Notmuch only

Is there an equivalent to the IMAP idle mode?

### Maildir only

**Comment by matthiasbeyer** [comment]
> Is there an equivalent to the IMAP idle mode?

No. notmuch works locally only. All you can do with notmuch (feel free to try out the notmuch commandline, it gives a good impression IMO), is _searching_ for emails and it gives you back a list of emails. It gives you back either a list of message IDs, or a list of mails plain-text and so on. So the "Common to all">"Connect" wouldn't even be part of notmuch in a sense that it "connects" to a remote server. That would just open the database actually.
Fetching and sending mail in notmuch land is done via other mechanisms, like offlineimap, mbsync or (in my case) via muchsync, which syncs notmuch databases.

**Comment by soywod** [comment]
> feel free to try out the notmuch commandline

I read the notmuch doc, I understand a bit better. I also plan to try the tool this week!

Here a bunch of questions:

- Notmuch creates its own database from a Maildir folder, right? So in fact the Maildir backend and the Notmuch backend are kind of complementary?
- How Notmuch treats mailboxes? As tags?

Here the updated table of interface:

|Method|IMAP|POP|Notmuch|Maildir|
|--|--|--|--|--|
|connect|login|login|[update the database](https://notmuchmail.org/manpages/notmuch-new-1/)|noop|
|read mailboxes|list mailboxes|list mailboxes|?|list sub-folders|
|read envelopes|fetch/search|list|[search](https://notmuchmail.org/manpages/notmuch-search-1/)|search files|
|read message|fetch|retrieve|[show](https://notmuchmail.org/manpages/notmuch-show-1/)|read file|
|add message|append|?|[insert](https://notmuchmail.org/manpages/notmuch-insert-1/)|add file|
|add flag|add flags|?|[flag +](https://notmuchmail.org/manpages/notmuch-tag-1/)|rename file|
|set flag|set flags|?|[flag + --remove-all](https://notmuchmail.org/manpages/notmuch-tag-1/)|rename file|
|remove flag|remove flags|?|[flag -](https://notmuchmail.org/manpages/notmuch-tag-1/)|rename file|
|disconnect|logout|logout|noop|noop|

**Comment by soywod** [comment]
> Notmuch does not know about "mailboxes" at all. It just gives you search over all your mails. What "show" or "search" returns, though, could be seen as a temporary mailbox!

I see. So in fact, it is sth to customize by the user!

> disconnect: Closing the database would be the notmuch operation here, but this is of course only an internal thing, nothing a user knows about.

Yes of course, but all this interface is internal. The public API will not include this.

> There is _some_ features that notmuch supports that are not represented here, most notably `notmuch-compact` for database reindexing and compacting.

This could be sth to do also in the `disconnect` method (controllable with an option)

Alright, thanks for these replies, I see more or less where to go. If you don't have any remark on https://github.com/soywod/himalaya/pull/296/files I will merge it and then start to develop the interface.

**Comment by soywod** [comment]
I almost finished the Maildir backend. I still need to adjust the trait, I noticed some difference between IMAP and Maildir that are not really compatible. Then we will be able to set up Notmuch!

**Comment by soywod** [comment]
It was more complicated than expected. First, I had hard time to figure out how to design the architecture (and also how to implement it, I did a lot of experimentations). Then, while implementing the Maildir backend with the `maildir` crate, I noticed that the lib missed some Maildir++ features, I needed to send a PR (which has been merged quickly https://github.com/staktrace/maildir/pull/27). Here so far the final `Backend` trait:

```rust
pub trait Backend<'a> {
    fn connect(&mut self) -> Result<()> {
        Ok(())
    }

    fn add_mbox(&mut self, mbox: &str) -> Result<()>;
    fn get_mboxes(&mut self) -> Result<Box<dyn Mboxes>>;
    fn del_mbox(&mut self, mbox: &str) -> Result<()>;
    fn get_envelopes(
        &mut self,
        mbox: &str,
        sort: &str,
        filter: &str,
        page_size: usize,
        page: usize,
    ) -> Result<Box<dyn Envelopes>>;
    fn add_msg(&mut self, mbox: &str, msg: &[u8], flags: &str) -> Result<Box<dyn ToString>>;
    fn get_msg(&mut self, mbox: &str, id: &str) -> Result<Msg>;
    fn copy_msg(&mut self, mbox_src: &str, mbox_dst: &str, ids: &str) -> Result<()>;
    fn move_msg(&mut self, mbox_src: &str, mbox_dst: &str, ids: &str) -> Result<()>;
    fn del_msg(&mut self, mbox: &str, ids: &str) -> Result<()>;
    fn add_flags(&mut self, mbox: &str, ids: &str, flags: &str) -> Result<()>;
    fn set_flags(&mut self, mbox: &str, ids: &str, flags: &str) -> Result<()>;
    fn del_flags(&mut self, mbox: &str, ids: &str, flags: &str) -> Result<()>;

    fn disconnect(&mut self) -> Result<()> {
        Ok(())
    }
}
```

`get_mboxes` and `get_envelopes` return a trait object, which allow us to have different mbox and envelope types for each backend implementation. However, `get_msg` is a concrete type, because no matter the backend it will always be parsed from the raw data.

Here the PR https://github.com/soywod/himalaya/pull/299, feel free to comment. I will merge it soon!

**Comment by soywod** [comment]
The Maildir feature has been merged on `master` :tada:, the next one is Notmuch :wink:

**Comment by matthiasbeyer** [comment]
It is basically the same as with any other library outside of the rust ecosystem (think openssl): It has to be installed on the target system. So when a user uses their package manager to install your software, the package must depend on openssl (or libnotmuch in this case).

**Comment by soywod** [comment]
I see. So it is sth to explain well in the doc. I will start the backend very soon and let you know!

**Comment by matthiasbeyer** [comment]
IMO Ids shouldn't be user visible. They should be used internally for searching for mails that a user flagged for example, for internal optimizations and such, but not be visible to users. A user does not have to know the ID of a message.

**Comment by soywod** [comment]
But it is a CLI, and an id is the only thing a user can get to select a specific email. But what you say joins what I think, internal ids should not be visible. A mapping is still necessary IMO, otherwise how can you select read/copy/move/delete a specific email in a CLI mode?

**Comment by matthiasbeyer** [comment]
yes, internally it is relevant. It can be a `pub struct MailId(String)`, tho, right? :smile:

**Comment by soywod** [comment]
So far what I have with the Notmuch backend:

- `get_mboxes` lists virtual mailboxes present in the config file (I introduced a new account entry `mailboxes: HashMap<String, String>` that are used for aliasing mailbox names, but it can be used as virtual mailboxes with the Notmuch backend by giving queries — for eg. `inbox = "tag:inbox"`)
- `get_envelopes` lists envelopes found in the given virtual mailbox (`inbox` by default)
- `find_envelopes` lists envelopes found in the given query
- `get_msg` reads the message found with the given id

I'm a bit stuck with the `xxx_flags` methods, because I'm not sure which method to call on the crate `notmuch-rs`. The most relevant one I found is https://docs.rs/notmuch/0.7.1/notmuch/struct.Query.html#method.add_tag_exclude, can you confirm?

**Comment by soywod** [comment]
> I didn't even know that notmuch has such an API

Yes, you have a `--format=json` to print a full JSON string (like himalaya in fact).

> This is the way. CI is cheap, developing against a JSON API is not

I agree. Thank you!

**Comment by soywod** [comment]
Do you think I should also put IMAP and Maildir as features (enabled by default)? This way, someone who wants to use Himalaya only with Maildir could have a smaller executable.

**Comment by soywod** [comment]
> Yes, I think it is a good idea.

I created a dedicated issue for this, it looks bigger as expected https://github.com/soywod/himalaya/issues/318.

I just released the `v0.5.7` with an experimental implementation of the Notmuch backend. For now the only way to have it is to compile himalaya sources with cargo and with the `--features notmuch`. I will push the new doc very soon.

**Comment by soywod** [comment]
All features have been merged into `master`, I just released the `v0.5.8`. Thank you for your precious help! If you find any bug or request feel free to ask.
